### ISSUE TITLE
update_column() utility not working as intended?

### ISSUE THREAD

**Issue by mxndrwgrdnr** [UDST/urbansim_templates#89]
I'm suspicious that the new `urbansim_templates.utils.update_column()` method is not functioning properly, or at least only functioning in certain settings, because it definitely _is_ working in some places, but not in others. I have a hypothesis as to why this might be happening, but I wonder what you @smmaurer think.

Here I am running the mode choice model in activitySynth:
![screen shot 2019-01-31 at 12 51 16 pm](https://user-images.githubusercontent.com/1395651/52084435-f3d8d600-2556-11e9-8f7c-a451c39fc054.png)

which uses the small MNL template and as such should automatically append the results to the correct output table as seen [here](https://github.com/UDST/urbansim_templates/blob/master/urbansim_templates/models/small_multinomial_logit.py#L385-L389).

But that doesn't happen:
![screen shot 2019-01-31 at 12 53 16 pm](https://user-images.githubusercontent.com/1395651/52084565-37cbdb00-2557-11e9-9892-443ad26b6df5.png)

In order to figure out if the problem might be in the model step itself, I manually trigger the `update_column()` command but to no avail:
![screen shot 2019-01-31 at 12 55 23 pm](https://user-images.githubusercontent.com/1395651/52084709-84afb180-2557-11e9-883d-1caca4525ef3.png)

So then I debug these [lines of code](https://github.com/UDST/urbansim_templates/blob/master/urbansim_templates/utils.py#L252-L255) from the `update_column()` method itself to see if the problem lies there, but it seems to work as intended:
![screen shot 2019-01-31 at 12 56 20 pm](https://user-images.githubusercontent.com/1395651/52084757-a315ad00-2557-11e9-997d-3ef631ae93f8.png)

But actually they don't because the updated column cannot be found in the parent table:
![screen shot 2019-01-31 at 12 58 38 pm](https://user-images.githubusercontent.com/1395651/52084897-f7b92800-2557-11e9-8b98-415e9dc9c64d.png)

So my working hypothesis is that for some reason, in some instances, updates that are made to an `orca.get_table()` dataframe wrapper object are not made to the orca table itself. I believe this has to do with the fact that successive `orca.get_table()` calls to the same table return objects that are stored in different locations in memory, as seen here:
![screen shot 2019-01-31 at 1 01 43 pm](https://user-images.githubusercontent.com/1395651/52085082-639b9080-2558-11e9-9a68-98d2a3f24b1a.png)

Whether or not these having different in-memory addresses is the issue, the following exercise shows how changes do not permeate:
![screen shot 2019-01-31 at 1 03 11 pm](https://user-images.githubusercontent.com/1395651/52085163-99d91000-2558-11e9-8928-f45d4d66d917.png)

Am I missing something here or is my reasoning sound? The bigger mystery to me now is why the `update_column()` method would _ever_ work to update the parent orca table itself.

**Comment by mxndrwgrdnr** [comment]
Here's a thought, could this have to do with whether or not the `cache=True` flag is set when the table is created?

**Comment by smmaurer** [comment]
Interesting (and troubling). Agree that it seems like `DataFrameWrapper.update_column()` is not reliably updating the underlying table in Orca.

I'm not sure the last part (beginning with "So my working hypothesis") is the reason, though -- i don't think i'd expect changes to propagate between copies anyway, even if they'd been recorded by Orca. 

I'm looking through the Orca documentation to see if i can come up with any leads..

**Comment by smmaurer** [comment]
It does seem like this is the intended workflow. And here's an example of the underlying data being updated properly, in the Large MNL unit tests: [test_large_multinomial_logit.py#L202-L211](https://github.com/UDST/urbansim_templates/blob/master/tests/test_large_multinomial_logit.py#L202-L211)

One thing to try might be explicitly replacing the orca table with the new one, which might be more reliable. So in your debugging section, something like:

```py
dfw = orca.get_table(mc.out_tables)
dfw.update_col(mc.out_column, mc.choices)
orca.add_table(mc.out_tables, dfw.to_frame())
```

If that works, it would be an easy fix to add it to the template utility.

**Comment by mxndrwgrdnr** [comment]
Hmm, and the `cache=True` flag isn't set in your tests so there goes my theory.

**Comment by mxndrwgrdnr** [comment]
I'll test the explicit table replacement and report back.

**Comment by bridwell** [comment]
@mxndrwgrdnr I think you're exactly right about the cache status. Basically what's going on here is (a) a function evaluates and returns a data frame, (b) you're modifying that data frame and then (c) the function is being evaluated again so you're obtaining the original data frame. 

```python
# define a non-cached table
@orca.table()
def my_table():
    return pd.DataFrame({'a': ['uno', 'dos', 'tres']})

# grab the table from orca and update it
my_wrapper = orca.get_table('my_table')
my_wrapper.update_col('b', pd.Series(['cuatro', 'cinco', 'seis']))

# grab the table from orca and update it
my_wrapper = orca.get_table('my_table')
my_wrapper.update_col('b', pd.Series(['cuatro', 'cinco', seis'])

# this will bring back the original result
orca.get_table('my_table').local
```
index | a 
-- | --
0|uno 
1|dos
2|tres

Instead if we cache, then the function is being re-evaluated and we get the same instance:
```python
@orca.table(cache=True, cache_scope='forever')
def my_table():
    return pd.DataFrame({'a': ['uno', 'dos', 'tres']})

my_wrapper= orca.get_table('my_table')
my_wrapper.update_col('b', pd.Series(['cuatro', 'cinco', 'seis']))
orca.get_table('my_table').local
```
index | a | b
-- | --|--
0|uno | cuatro
1|dos | cinco
2|tres | seis

I lieu of messing with the cache, I suppose you could retrieve the data frame from the function and then add it back into orca as a data frame rather than a function. 

```python
@orca.table()
def my_table():
    return pd.DataFrame({'a': ['uno', 'dos', 'tres']})

orca.add_table('my_table', orca.get_table('my_table').local)
my_wrapper= orca.get_table('my_table')
my_wrapper.update_col('b', pd.Series(['cuatro', 'cinco', 'seis']))
orca.get_table('my_table').local
```
index | a | b
-- | --|--
0|uno | cuatro
1|dos | cinco
2|tres | seis

**Comment by mxndrwgrdnr** [comment]
OK, so then why do your large MNL unit tests work, because you aren't setting the cache [here](https://github.com/UDST/urbansim_templates/blob/master/tests/test_large_multinomial_logit.py#L23)

**Comment by bridwell** [comment]
That looks to be creating a orca table wrapper from an already created pandas.DataFrame, not a function that returns a pandas.DataFrame.

**Comment by mxndrwgrdnr** [comment]
Yes but there is a function that returns a pandas.DataFrame that gets called behind the scenes [here](https://github.com/UDST/urbansim_templates/blob/master/tests/test_large_multinomial_logit.py#L208)

**Comment by bridwell** [comment]
How is the __ __persons_CHTS_format__ __ table being added to orca? 

Like this? 
```python
#  create a data frame and add it explicitly
df = pandas.DataFrame({a: [1,2,3]})
orca.add_table('__persons_CHTS_format__', df)
```
Or like this?
```python
# wrap a function and add it to orca
@orca.table()
def __persons_CHTS_format__():
    return pandas.DataFrame({a: [1,2,3]})
```
If table is added as a wrapped function (2nd case), I would set cache=True_ and cache_scope_='forever' or add the table explicitly. If that doesn't fix it then there's a template or orca problem.

**Comment by mxndrwgrdnr** [comment]
@bridwell I believe the default is `scope=forever`, so no worries there.

But yeah, @smmaurer I think this solution (1-3) makes the most sense. 

Regarding point 4, though, would the idea be to convert all of the `@orca.table()` calls and table func definitions in datasources.py files to a bunch of separate `pd.read_csv()` and `orca.add_table()` calls? Would these get automatically injected into the orca namespace the same way when doing `from scripts import variables`?
